README.md
# Django Sage Session

## What is django-sage-session package?

django-sage-session is a useful package for per user session & private session in Django Web Framework

##### The Latest version of [django-sage-session](https://django-sage-session.readthedocs.io/) documentation

![SageTeam](https://github.com/sageteam-org/django-sage-painless/blob/develop/docs/images/tag_sage.png?raw=true "SageTeam")

![PyPI release](https://img.shields.io/pypi/v/django-sage-session "django-sage-session")
![Supported Python versions](https://img.shields.io/pypi/pyversions/django-sage-session "django-sage-session")
![Supported Django versions](https://img.shields.io/pypi/djversions/django-sage-session "django-sage-session")
![Documentation](https://img.shields.io/readthedocs/django-sage-session "django-sage-session")

- [Project Detail](#project-detail)
- [Git Rules](#git-rules)
- [Get Started](#get-started)
- [Usage](#usage)
- [Settings](#settings)
- [Admin](#admin)
## Project Detail

- Language: Python > 3.5
- Framework: Django > 3.1

## Git Rules

S.A.G.E. team Git Rules Policy is available here:

- [S.A.G.E. Git Policy](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow)

## Get Started

First install the package using pip:

```shell
$ pip install django-sage-session
```

Then add sage_cache to `INSTALLED_APPS`:

```python
INSTALLED_APPS = [
    ...
    'sage_session',
    ...
]
```

## Usage

For using per user session you have to modify `MIDDLEWARE` in setting:

Replace django Session middleware with sage_session User Session middleware

```python
MIDDLEWARE = [
    ...
    'django.middleware.security.SecurityMiddleware',
    'sage_session.middlewares.UserSessionMiddleware', # provided middleware for per user session
    # 'django.contrib.sessions.middleware.SessionMiddleware',
    ...
]
```

For activating Privacy session add it to `MIDDLEWARE`:

Add after UserSessionMiddleware

```python
MIDDLEWARE = [
    ...
    'sage_session.middlewares.UserSessionMiddleware',  # user session
    'sage_session.middlewares.UserSessionPrivacyMiddleware',  # privacy session
    # 'django.contrib.sessions.middleware.SessionMiddleware',
    ...
]
```

## Settings
```python
PRIVACYSESSIONS_AUTHED_ONLY = False  # authentication required 
PRIVACYSESSIONS_PRIVACY_IP = True  # validate IP
PRIVACYSESSIONS_PRIVACY_UA = True  # validate user agent
PRIVACYSESSIONS_REMOTE_ADDR_KEY = 'REMOTE_ADDR'  # remote addr key in request (not recommended to change)
PRIVACYSESSIONS_REDIRECT_VIEW = None  # redirect view after logging out
PRIVACYSESSIONS_FAILURE_STATUS = 400 # if not PRIVACYSESSIONS_REDIRECT_VIEW set return status code
PRIVACYSESSIONS_IPV4_LENGTH = 32  # length of ipv4 (not recommended to change)
PRIVACYSESSIONS_IPV6_LENGTH = 64  # length of ipv6 (not recommended to change)
PRIVACYSESSIONS_IP_KEY = '_privacysessions_ip'  # ip key in session
PRIVACYSESSIONS_UA_KEY = '_privacysessions_ua'  # user agent key in session
```

## Admin

`sage_session` also has an admin panel for User Session monitoring, It will activate when you add `sage_session` to `INSTALLED_APPS`
## Team
| [<img src="https://github.com/sageteam-org/django-sage-painless/blob/develop/docs/images/sepehr.jpeg?raw=true" width="230px" height="230px" alt="Sepehr Akbarzadeh">](https://github.com/sepehr-akbarzadeh) | [<img src="https://github.com/sageteam-org/django-sage-painless/blob/develop/docs/images/mehran.png?raw=true" width="225px" height="340px" alt="Mehran Rahmanzadeh">](https://github.com/mrhnz) |
| ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- |
| [Sepehr Akbarazadeh Maintainer](https://github.com/sepehr-akbarzadeh)                                                                                                             | [Mehran Rahmanzadeh Maintainer](https://github.com/mrhnz)                                                                                                       |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.1.0] - 2021-07-30
### Added
- Created project

[0.1.0]: https://github.com/sageteam-org/django-sage-session/commits/develop

sage_session/db.py
import logging

from django.contrib import auth
from django.contrib.sessions.backends.base import CreateError, SessionBase
from django.contrib.sessions.backends.db import SessionStore
from django.core.exceptions import SuspiciousOperation
from django.db import IntegrityError, router, transaction
from django.utils import timezone
from django.utils.encoding import force_text


class UserSessionStore(SessionStore):
    """Implements database session store."""

    def __init__(
            self,
            session_key=None,
            user_agent=None,
            ip=None,
            location=None,
            device_brand=None,
            device_model=None,
            browser=None,
            browser_version=None,
            os=None,
            os_version=None,
            is_mobile=False,
            is_tablet=False,
            is_touch_capable=False,
            is_pc=False,
            is_bot=False
    ):
        super(UserSessionStore, self).__init__(session_key)
        # Truncate user_agent string to max_length of the CharField
        self.user_agent = user_agent[:200] if user_agent else user_agent
        self.ip = ip
        self.location = location
        self.device_model = device_model
        self.device_brand = device_brand
        self.browser = browser
        self.browser_version = browser_version
        self.os = os
        self.os_version = os_version
        self.is_mobile = is_mobile
        self.is_tablet = is_tablet
        self.is_touch_capable = is_touch_capable
        self.is_pc = is_pc
        self.is_bot = is_bot
        self.user_id = None

    def __setitem__(self, key, value):
        if key == auth.SESSION_KEY:
            self.user_id = value
        super(UserSessionStore, self).__setitem__(key, value)

    def load(self):
        try:
            s = Session.objects.get(
                session_key=self.session_key,
                expire_date__gt=timezone.now()
            )
            self.user_id = s.user_id
            return self.decode(s.session_data)
        except (Session.DoesNotExist, SuspiciousOperation) as e:
            if isinstance(e, SuspiciousOperation):
                logger = logging.getLogger(
                    'django.security.%s' % e.__class__.__name__
                )
                logger.warning(force_text(e))
            return {}

    def save(self, must_create=False):
        """Saves the current session data to the database. If 'must_create' is
        True, a database error will be raised if the saving operation doesn't
        create a *new* entry (as opposed to possibly updating an existing
        entry).
        """
        if self.session_key is None:
            return self.create()
        data = self._get_session(no_load=must_create)
        obj = Session(
            session_key=self._get_or_create_session_key(),
            session_data=self.encode(data),
            expire_date=self.get_expiry_date(),
            user_agent=self.user_agent,
            user_id=self.user_id,
            ip=self.ip,
            location=self.location,
            device_brand=self.device_brand,
            device_model=self.device_model,
            browser=self.browser,
            browser_version=self.browser_version,
            os=self.os,
            os_version=self.os_version,
            is_mobile=self.is_mobile,
            is_tablet=self.is_tablet,
            is_touch_capable=self.is_touch_capable,
            is_pc=self.is_pc,
            is_bot=self.is_bot
        )
        using = router.db_for_write(Session, instance=obj)
        try:
            with transaction.atomic(using):
                self.delete_duplicate()
                obj.save(force_insert=must_create, using=using)
        except IntegrityError as e:
            if must_create and 'session_key' in str(e):
                raise CreateError
            raise

    def delete_duplicate(self):
        """delete other session objects for user"""
        return Session.objects.filter(user_id=self.user_id).delete()

    def clear(self):
        """clear session"""
        super(UserSessionStore, self).clear()
        self.user_id = None


from sage_session.models import Session


sage_session/admin.py
from django.contrib import admin
from django.utils.timezone import now
from django.utils.translation import ugettext_lazy as _

from sage_session.models import Session


@admin.register(Session)
class SessionAdmin(admin.ModelAdmin):
    """Session admin"""
    list_display = (
        'user', 'ip', 'location',
        'device_brand', 'device_model', 'browser',
        'os', 'is_mobile', 'is_tablet',
        'is_touch_capable', 'is_pc', 'is_valid',
    )
    search_fields = (
        'user__username',
        'user__first_name',
        'user__last_name',
    )
    readonly_fields = ('user', 'session_key')
    fieldsets = (
        (_('Security'), {
            'fields': (
                'session_key',
                'session_data',
                'expire_date',
            )
        }),
        (_('Info'), {
            'fields': (
                'user',
                'user_agent',
                'ip',
                'location',
                'device_brand',
                'device_model',
                'browser',
                'browser_version',
                'os',
                'os_version',
                'is_mobile',
                'is_tablet',
                'is_touch_capable',
                'is_pc',
                'is_bot'
            )
        })
    )

    def is_valid(self, obj):
        return obj.expire_date > now()

    is_valid.boolean = True


sage_session/models.py
from django.db import models
from django.conf import settings
from django.contrib.sessions.models import (
    AbstractBaseSession,
    BaseSessionManager
)
from django.utils.translation import ugettext_lazy as _


class SessionManager(BaseSessionManager):
    """Session manager"""
    use_in_migrations = True

    def encode(self, session_dict):
        """Returns the given session dictionary serialized and encoded as a string."""
        return UserSessionStore().encode(session_dict)

    def save(self, session_key, session_dict, expire_date):
        s = self.model(session_key, self.encode(session_dict), expire_date)
        if session_dict:
            s.save()
        else:
            s.delete()  # Clear sessions with no data.
        return s


class Session(AbstractBaseSession):
    """Session model
    containing user session information
    Additionally this session object provides the following properties:
    `user`, `user_agent`, `location`, `device`, `last_activity` and `ip`.
    """
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        null=True,
        on_delete=models.CASCADE,
        verbose_name=_('User')
    )
    user_agent = models.CharField(
        _('User Agent'),
        null=True,
        blank=True,
        max_length=200
    )
    last_activity = models.DateTimeField(
        _('Last Activity'),
        auto_now=True
    )
    ip = models.GenericIPAddressField(
        null=True,
        blank=True,
        verbose_name=_('IP')
    )
    location = models.CharField(
        _('Location'),
        max_length=50,
        null=True,
        blank=True
    )
    device_brand = models.CharField(
        _('Device Brand'),
        max_length=50,
        null=True,
        blank=True
    )
    device_model = models.CharField(
        _('Device Model'),
        max_length=50,
        null=True,
        blank=True
    )
    browser = models.CharField(
        _('Browser'),
        max_length=50,
        null=True,
        blank=True
    )
    browser_version = models.CharField(
        _('Browser Version'),
        max_length=50,
        null=True,
        blank=True
    )
    os = models.CharField(
        _('OS'),
        max_length=50,
        null=True,
        blank=True
    )
    os_version = models.CharField(
        _('OS Version'),
        max_length=50,
        null=True,
        blank=True
    )
    is_mobile = models.BooleanField(
        _('Is mobile'),
        default=False
    )
    is_tablet = models.BooleanField(
        _('Is tablet'),
        default=False
    )
    is_touch_capable = models.BooleanField(
        _('Is touch capable'),
        default=False
    )
    is_pc = models.BooleanField(
        _('Is pc'),
        default=False
    )
    is_bot = models.BooleanField(
        _('Is bot'),
        default=False
    )

    objects = SessionManager()

    def __str__(self):
        return f'Session for user {self.user}'


from sage_session.db import UserSessionStore


sage_session/services.py
from django.contrib.gis.geoip2 import GeoIP2

from user_agents import parse


def analyse_user_agent(user_agent):
    """Analyse user agent string and extract information
    :param user_agent: user agent string
    :type user_agent: str
    :return: information about user device and etc
    :rtype: tuple (
        device brand,
        device model,
        browser,
        browser version,
        os,
        os version,
        is mobile,
        is tablet,
        is touch capable,
        is pc,
        is bot
    )
    """
    agent = parse(user_agent)
    return (
        agent.device.brand,
        agent.device.model,
        agent.browser.family,
        agent.browser.version_string,
        agent.os.family,
        agent.os.version_string,
        agent.is_mobile,
        agent.is_tablet,
        agent.is_touch_capable,
        agent.is_pc,
        agent.is_bot
    )


def get_ip_location(ip):
    """Get location information of ip
    :param ip: IP address
    :type ip: str
    :return: city & country of the ip address
    :rtype: tuple (country, city)
    """
    g = GeoIP2()
    data = g.city(ip)
    return (
        data.get('country_name'),
        data.get('city')
    )


sage_session/settings.py
from django.conf import settings

PRIVACYSESSIONS_AUTHED_ONLY = getattr(settings, 'PRIVACYSESSIONS_AUTHED_ONLY', False)
PRIVACYSESSIONS_REMOTE_ADDR_KEY = getattr(settings, 'PRIVACYSESSIONS_REMOTE_ADDR_KEY', 'REMOTE_ADDR')
PRIVACYSESSIONS_REDIRECT_VIEW = getattr(settings, 'PRIVACYSESSIONS_REDIRECT_VIEW', None)
PRIVACYSESSIONS_FAILURE_STATUS = getattr(settings, 'PRIVACYSESSIONS_FAILURE_STATUS', 400)
PRIVACYSESSIONS_PRIVACY_IP = getattr(settings, 'PRIVACYSESSIONS_PRIVACY_IP', True)
PRIVACYSESSIONS_IPV4_LENGTH = getattr(settings, 'PRIVACYSESSIONS_IPV4_LENGTH', 32)
PRIVACYSESSIONS_IPV6_LENGTH = getattr(settings, 'PRIVACYSESSIONS_IPV6_LENGTH', 64)
PRIVACYSESSIONS_PRIVACY_UA = getattr(settings, 'PRIVACYSESSIONS_PRIVACY_UA', True)
PRIVACYSESSIONS_IP_KEY = getattr(settings, 'PRIVACYSESSIONS_IP_KEY', '_privacysessions_ip')
PRIVACYSESSIONS_UA_KEY = getattr(settings, 'PRIVACYSESSIONS_UA_KEY', '_privacysessions_ua')


sage_session/middlewares.py
import warnings
import logging

from netaddr import (
    IPNetwork,
    IPAddress,
    AddrConversionError,
    AddrFormatError
)

from django.contrib.sessions.middleware import SessionMiddleware
from django.utils.deprecation import MiddlewareMixin
from django.http import HttpResponse
from django.conf import settings as dj_settings
from django.urls import reverse
from django.shortcuts import redirect
from django.utils.encoding import force_text

from sage_session.db import UserSessionStore
from sage_session.services import analyse_user_agent, get_ip_location
from sage_session import settings

SESSION_IP_KEY = getattr(settings, 'PRIVACYSESSIONS_IP_KEY', '_privacysessions_ip')
SESSION_UA_KEY = getattr(settings, 'PRIVACYSESSIONS_UA_KEY', '_privacysessions_ua')

logger = logging.getLogger('privacysessions')


class UserSessionMiddleware(SessionMiddleware):
    """A middleware that adds ip and user_agent to the session store."""

    def process_request(self, request):
        session_key = request.COOKIES.get(dj_settings.SESSION_COOKIE_NAME, None)
        ip = request.META.get('REMOTE_ADDR', '')
        try:
            location = get_ip_location(ip)
        except Exception as e:
            warnings.warn(str(e))
            location = None
        user_agent = request.META.get('HTTP_USER_AGENT', '')
        (
            device_brand,
            device_model,
            browser,
            browser_version,
            os,
            os_version,
            is_mobile,
            is_tablet,
            is_touch_capable,
            is_pc,
            is_bot
        ) = analyse_user_agent(user_agent)
        request.session = UserSessionStore(
            ip=ip,
            user_agent=user_agent,
            session_key=session_key,
            location=location,
            device_model=device_model,
            device_brand=device_brand,
            browser=browser,
            browser_version=browser_version,
            os=os,
            os_version=os_version,
            is_mobile=is_mobile,
            is_tablet=is_tablet,
            is_touch_capable=is_touch_capable,
            is_pc=is_pc,
            is_bot=is_bot
        )


class UserSessionPrivacyMiddleware(MiddlewareMixin):
    def process_request(self, request):
        # Short circuit when request doesn't have session
        if not hasattr(request, 'session'):
            return

        # Short circuit for option to require authed users
        if getattr(settings, 'PRIVACYSESSIONS_AUTHED_ONLY', False):
            user = getattr(request, 'user', None)
            # No logged in user -- ignore checks
            if not user or not hasattr(user, 'is_authenticated') or not user.is_authenticated:
                return

        # Extract remote IP address for validation purposes
        remote_addr_key = getattr(settings, 'PRIVACYSESSIONS_REMOTE_ADDR_KEY', 'REMOTE_ADDR')
        remote_addr = request.META.get(remote_addr_key)
        if remote_addr:
            remote_addr = remote_addr.split(',')[0]

        # Clear the session and handle response when request doesn't validate
        if not self.validate_ip(request, remote_addr) or not self.validate_ua(request):
            if getattr(settings, 'PRIVACYSESSIONS_AUTHED_ONLY', False):
                from django.contrib.auth import logout
                logout(request)
            else:  # logout(...) flushes the session so ensure it only happens once
                request.session.flush()
            logger.warning("Destroyed session due to invalid change of remote host or user agent")
            redirect_view = getattr(settings, 'PRIVACYSESSIONS_REDIRECT_VIEW', None)
            if redirect_view:
                return redirect(reverse(redirect_view))
            else:
                status = getattr(settings, 'PRIVACYSESSIONS_FAILURE_STATUS', 400)
                return HttpResponse(status=status)

        # Set the UA/IP Address on the session since they validated correctly
        request.session[SESSION_IP_KEY] = remote_addr
        if request.META.get('HTTP_USER_AGENT'):
            request.session[SESSION_UA_KEY] = force_text(request.META['HTTP_USER_AGENT'], errors='replace')

    def validate_ip(self, request, remote_ip):
        """check request ip and session stored ip"""
        # When we aren't configured to restrict on IP address
        if not getattr(settings, 'PRIVACYSESSIONS_PRIVACY_IP', True):
            return True
        # When the IP address key hasn't yet been set on the request session
        if SESSION_IP_KEY not in request.session:
            return True
        # When there is no remote IP, check if one has been set on the session
        session_ip = request.session[SESSION_IP_KEY]
        if not remote_ip:
            if session_ip:  # session has remote IP value so validate :-(
                return False
            else:  # Session doesn't have remote IP value so possibly :-)
                return True

        # Compute fuzzy IP compare based on settings on compare sensitivity
        if not session_ip:
            return False
        session_network = IPNetwork(session_ip)
        remote_ip = IPAddress(remote_ip)
        try:
            session_network = session_network.ipv4()
            remote_ip = remote_ip.ipv4()
            session_network.prefixlen = getattr(settings, 'PRIVACYSESSIONS_IPV4_LENGTH', 32)
        except AddrConversionError:
            try:
                session_network.prefixlen = getattr(settings, 'PRIVACYSESSIONS_IPV6_LENGTH', 64)
            except AddrFormatError:
                # session_network must be IPv4, but remote_ip is IPv6
                return False
        return remote_ip in session_network

    def validate_ua(self, request):
        """check request user agent and session stored user agent"""
        # When we aren't configured to restrict on user agent
        if not getattr(settings, 'PRIVACYSESSIONS_PRIVACY_UA', True):
            return True
        # When the user agent key hasn't been set yet in the request session
        if SESSION_UA_KEY not in request.session:
            return True
        # Compare the new user agent value with what is known about the session
        return request.session[SESSION_UA_KEY] == force_text(request.META['HTTP_USER_AGENT'], errors='replace')
